open-hackathon-server/src/hackathon/hack/host_server_manager.py

```diff
@@ -190,6 +209,142 @@ def create_docker_host_vm(self, hackathon_id):
+                    vm_data["container_count"] = len(jsonObject)
+        except Exception as e:
+            self.log.error(e)
+
+        return vm_data
+
+    def update_host_server(self, hackathon_id, args):
+        """
+        update a docker host's information for a hackathon.
+
+        :param hackathon_id: the id of hackathon in DB
+        :type hackathon_id: Integer
+
+        :return: ok() if succeed.
+                 precondition_failed(...) if fail to update the docker_host's information
+        """
+        vm = self.db.find_first_object_by(DockerHostServer, id=args.id, hackathon_id=hackathon_id)
+        if vm is None:
+            self.log.warn('delete docker_host fail, not find hostserver_id:' + args.id
+                          + 'hackathon_id:' + hackathon_id)
+            return precondition_failed("", "host_server not found")
+
+        try:
+            ping_url = 'http://%s:%d/_ping' % (args.public_dns, int(args.public_docker_api_port))
+            req = requests.get(ping_url, timeout=5)
+            if req.status_code == 200 and req.content == DockerPingResult.OK:
+                args['state'] = DockerHostServerStatus.DOCKER_READY
+            else:
+                args['state'] = DockerHostServerStatus.UNAVAILABLE
+        except Exception as e:
+            self.log.error(e)
+            args['state'] = DockerHostServerStatus.UNAVAILABLE
+
+        self.db.update_object(vm, vm_name = args.vm_name,
+                                  public_dns = args.public_dns,
+                                  public_ip = args.public_ip,
+                                  public_docker_api_port = args.public_docker_api_port,
+                                  private_ip = args.private_ip,
+                                  private_docker_api_port = args.private_docker_api_port,
+                                  container_max_count = args.container_max_count,
+                                  state = args.state,
+                                  disabled = args.disabled,
+                                  update_time = self.util.get_now())
+        return ok()
+
+    def delete_host_server(self, hackathon_id, host_server_id):
+        """
+        delete a docker host for a hackathon.
+
+        :param hackathon_id: the id of a hackathon in DB
+        :type hackathon_id: Integer
+
+        :param host_server_id: the id of host_server in DB
+        :type host_server_id: Integer
+
+        :return: ok() if succeeds.
+                 precondition_failed(...) if fail to delete the docker_host
+        """
+        vm = self.db.find_first_object_by(DockerHostServer, id=host_server_id, hackathon_id=hackathon_id)
+        if vm is None:
+            self.log.warn('delete docker_host fail, not find hostserver_id:' + host_server_id
+                          + 'hackathon_id:' + hackathon_id)
+            return precondition_failed("", "host_server not found")
```

Comment:
You cannot prevent user from deleting the same object twice or more. Just return `ok()` in case caller want to delete objects that are not existed. In fact, there is a specific http code `204`(`No content`) which indicates success in deleting something no matter it's deleted this time or before. `204` is kind of `success` and we use `ok`,`200` instead for simplicity in frontend.